Incorrect redirecting assembly to the GAC

Hello,

I'm try to use **ConfuserEx 1.3.1** and I found that it makes redirect to an assembly from the GAC, when the really used assembly is located in the one of `probPath`. In my case such behavior produce an error (because the assembly from the GAC doesn't contain a couple of types).

```
 [INFO] Confuser.Core 1.3.1+cf4240c129 Copyright © 2014 Ki, 2018 Martin Karing
 [INFO] Running on Microsoft Windows NT 6.2.9200.0, .NET Framework v4.0.30319.42000, 64 bits
[DEBUG] Discovering plugins...
 [INFO] Discovered 10 protections, 1 packers.
[DEBUG] Resolving component dependency...
 [INFO] Loading input modules...
 [INFO] Loading 'TestConfuserEx.dll'...
 [INFO] Initializing...
[DEBUG] Building pipeline...
 [INFO] Resolving dependencies...
[DEBUG] Checking Strong Name...
[DEBUG] Creating global .cctors...
[DEBUG] Watermarking...
[DEBUG] Executing 'Name analysis' phase...
[DEBUG] Building VTables & identifier list...
[ERROR] Failed to resolve a type, check if all dependencies are present in the correct version.
Exception: dnlib.DotNet.TypeResolveException: Could not resolve type: Microsoft.Build.Framework.IBuildEngine5 (Microsoft.Build.Framework, Version=15.1.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a)
   at dnlib.DotNet.TypeRef.ResolveThrow(ModuleDef sourceModule)
   at Confuser.Renamer.VTableStorage.GetVTable(ITypeDefOrRef type)
   at Confuser.Renamer.VTable.ConstructVTable(TypeDef typeDef, VTableStorage storage)
   at Confuser.Renamer.VTableStorage.GetOrConstruct(TypeDef type)
   at Confuser.Renamer.AnalyzePhase.Execute(ConfuserContext context, ProtectionParameters parameters)
   at Confuser.Core.ProtectionPipeline.ExecuteStage(PipelineStage stage, Action`1 func, Func`1 targets, ConfuserContext context)
   at Confuser.Core.ConfuserEngine.RunPipeline(ProtectionPipeline pipeline, ConfuserContext context)
   at Confuser.Core.ConfuserEngine.RunInternal(ConfuserParameters parameters, CancellationToken token)
[ERROR] ---BEGIN DEBUG INFO---
[ERROR] Installed Framework Versions:
[ERROR]     v4
[ERROR]     Client  4.8.03761
[ERROR]     Full  4.8.03761
[ERROR]     v4.0
[ERROR]     Client  4.0.0.0
[ERROR]     v4.5 528049
[ERROR] 
[ERROR] Cached assemblies:
[ERROR]     TestConfuserEx, Version=1.0.0.0, Culture=neutral, PublicKeyToken=null (C:\Users\GAlexandriya\source\repos\TestConfuserEx\TestConfuserEx\bin\Release\TestConfuserEx.dll)
[ERROR]         mscorlib, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089
[ERROR]         Microsoft.Build.Framework, Version=15.1.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a
[ERROR]     mscorlib, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089 (C:\WINDOWS\Microsoft.NET\assembly\GAC_32\mscorlib\v4.0_4.0.0.0__b77a5c561934e089\mscorlib.dll)
[ERROR]     Microsoft.Build.Framework, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a (C:\WINDOWS\Microsoft.NET\assembly\GAC_MSIL\Microsoft.Build.Framework\v4.0_4.0.0.0__b03f5f7f11d50a3a\Microsoft.Build.Framework.dll)
[ERROR]         mscorlib, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089
[ERROR]         System.Xaml, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089
[ERROR]         System, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089
[ERROR] ---END DEBUG INFO---
Failed at 3:43 PM, 0:00 elapsed.
``` 
Here you can see when redirection was applied in the dnlib: 
<details>
<summary>Click to expand pictures</summary>

![ConfuserEx_1](https://user-images.githubusercontent.com/13402478/82815086-148e1300-9ea1-11ea-950e-5825c569fb63.png)

![ConfuserEx_2](https://user-images.githubusercontent.com/13402478/82815085-13f57c80-9ea1-11ea-9e2e-4b6c0adbd2d7.png)

![image](https://user-images.githubusercontent.com/13402478/82815051-050eca00-9ea1-11ea-886c-f4926b61232f.png)

</details>

And [here](https://github.com/mkaring/ConfuserEx/blob/release/2.0/Confuser.Core/ConfuserEngine.cs#L149) ConfuserEx creates `AssemblyResolver`.

It is a small project to reproduce: [TestConfuserEx.zip](https://github.com/mkaring/ConfuserEx/files/4677266/TestConfuserEx.zip)

I didn't find any params to avoid this behavior (cannot remove assemblies from the GAC) and to allow ConfuserEx gets assemblies from `probPath`. Did I miss them? If no, do you plan to add them or apply any other kind of patch?

--- a/Confuser.Core/ConfuserAssemblyResolver.cs
+++ b/Confuser.Core/ConfuserAssemblyResolver.cs
@@ -1,54 +1,114 @@
 ﻿using System;
+using System.Collections;
 using System.Collections.Generic;
 using System.Linq;
 using dnlib.DotNet;
 
 namespace Confuser.Core {
 	internal sealed class ConfuserAssemblyResolver : IAssemblyResolver {
-		AssemblyResolver InternalResolver { get; }
+		internal AssemblyResolver InternalFuzzyResolver { get; } = new AssemblyResolver { FindExactMatch = false };
+		internal AssemblyResolver InternalExactResolver { get; } = new AssemblyResolver { FindExactMatch = true };
 
 		public bool EnableTypeDefCache {
-			get => InternalResolver.EnableTypeDefCache; 
-			set => InternalResolver.EnableTypeDefCache = value;
+			get => InternalFuzzyResolver.EnableTypeDefCache;
+			set {
+				InternalFuzzyResolver.EnableTypeDefCache = value;
+				InternalExactResolver.EnableTypeDefCache = value;
+			}
 		}
 
 		public ModuleContext DefaultModuleContext {
-			get => InternalResolver.DefaultModuleContext; 
-			set => InternalResolver.DefaultModuleContext = value;
+			get => InternalFuzzyResolver.DefaultModuleContext;
+			set {
+				InternalFuzzyResolver.DefaultModuleContext = value;
+				InternalExactResolver.DefaultModuleContext = value;
+			}
 		}
 
-		public IList<string> PostSearchPaths => InternalResolver.PostSearchPaths;
-
-		internal ConfuserAssemblyResolver() => 
-			InternalResolver = new AssemblyResolver();
+		public IList<string> PostSearchPaths => new TeeList(InternalFuzzyResolver.PostSearchPaths, InternalExactResolver.PostSearchPaths);
+		public IList<string> PreSearchPaths => new TeeList(InternalFuzzyResolver.PreSearchPaths, InternalExactResolver.PreSearchPaths);
 
 		/// <inheritdoc />
 		public AssemblyDef Resolve(IAssembly assembly, ModuleDef sourceModule) {
 			if (assembly is AssemblyDef assemblyDef)
 				return assemblyDef;
 
-			var cachedAssemblyDef = InternalResolver
-				.GetCachedAssemblies()
-				.FirstOrDefault(a => AssemblyNameComparer.NameAndPublicKeyTokenOnly.Equals(a, assembly));
-			if (!(cachedAssemblyDef is null))
-				return cachedAssemblyDef;
-
-			AssemblyDef resolvedAssemblyDef = null;
-			try {
-				InternalResolver.FindExactMatch = true;
-				resolvedAssemblyDef = InternalResolver.Resolve(assembly, sourceModule);
-			}
-			finally {
-				InternalResolver.FindExactMatch = false;
-			}
-
-			return resolvedAssemblyDef ?? InternalResolver.Resolve(assembly, sourceModule);
+			var resolvedAssemblyDef = InternalExactResolver.Resolve(assembly, sourceModule);
+			return resolvedAssemblyDef ?? InternalFuzzyResolver.Resolve(assembly, sourceModule);
 		}
 
-		public void Clear() => InternalResolver.Clear();
+		public void Clear() {
+			InternalExactResolver.Clear();
+			InternalFuzzyResolver.Clear();
+		}
 
-		public IEnumerable<AssemblyDef> GetCachedAssemblies() => InternalResolver.GetCachedAssemblies();
+		public IEnumerable<AssemblyDef> GetCachedAssemblies() => 
+			InternalExactResolver.GetCachedAssemblies().Concat(InternalFuzzyResolver.GetCachedAssemblies());
 
-		public void AddToCache(ModuleDefMD modDef) => InternalResolver.AddToCache(modDef);
+		public void AddToCache(ModuleDefMD modDef) {
+			InternalExactResolver.AddToCache(modDef);
+			InternalFuzzyResolver.AddToCache(modDef);
+		}
+
+		private sealed class TeeList : IList<string> {
+			private readonly IList<IList<string>> _lists;
+
+			internal TeeList(params IList<string>[] lists) => _lists = lists;
+
+			/// <inheritdoc />
+			public IEnumerator<string> GetEnumerator() => _lists[0].GetEnumerator();
+
+			/// <inheritdoc />
+			IEnumerator IEnumerable.GetEnumerator() => GetEnumerator();
+
+			/// <inheritdoc />
+			public void Add(string item) {
+				foreach (var list in _lists) 
+					list.Add(item);
+			}
+
+			/// <inheritdoc />
+			public void Clear() {
+				foreach (var list in _lists) 
+					list.Clear();
+			}
+
+			/// <inheritdoc />
+			public bool Contains(string item) => _lists[0].Contains(item);
+
+			/// <inheritdoc />
+			public void CopyTo(string[] array, int arrayIndex) => _lists[0].CopyTo(array, arrayIndex);
+
+			/// <inheritdoc />
+			public bool Remove(string item) =>
+				_lists.Aggregate(true, (current, list) => current | list.Remove(item));
+
+			/// <inheritdoc />
+			public int Count => _lists[0].Count;
+
+			/// <inheritdoc />
+			public bool IsReadOnly => _lists[0].IsReadOnly;
+
+			/// <inheritdoc />
+			public int IndexOf(string item) => _lists[0].IndexOf(item);
+
+			/// <inheritdoc />
+			public void Insert(int index, string item) {
+				foreach (var list in _lists) 
+					list.Insert(index, item);
+			}
+
+			/// <inheritdoc />
+			public void RemoveAt(int index) {
+				foreach (var list in _lists) 
+					list.RemoveAt(index);
+			}
+
+			/// <inheritdoc />
+			public string this[int index] {
+				get => _lists[0][index];
+				set => _lists[0][index] = value;
+			}
+		}
 	}
 }

